server.js

```diff
@@ -1,20 +1,125 @@
 const express = require("express");
 const cors = require("cors");
+const uuid = require("uuid");
+const moment = require("moment");
+const validator = require("email-validator");
 
 const app = express();
 
 app.use(express.json());
+app.use(express.urlencoded({ extended: true}));
 app.use(cors());
 
 //Use this array as your (in-memory) data store.
 const bookings = require("./bookings.json");
 
+
+function isInvalidId(id, index, response) {
+  if (index < 0) {
+    response.status(400).send("No booking with Id: " + id + " is found");
+  }
+}
+
+
 app.get("/", function (request, response) {
   response.send("Hotel booking server.  Ask for /bookings, etc.");
 });
 
 // TODO add your routes and helper functions here
 
+// GET All Bookings
+
+app.get("/bookings", function(req, res) {
+  if(bookings.length <= 0) {
+    res.status(500).send("No Available Data");
+  }
+  res.json(bookings)
+});
+
+// Search Bookings
+app.get("/bookings/search", function (req, res) {
+  let date = req.query.date;
+  let term = req.query.term;
+
+  let fltBookings = bookings
+    .filter(
+      (elt) =>
+        !date ||
+        (moment(elt.checkInDate) <= moment(date) &&
+          moment(elt.checkOutDate) >= moment(date))
+    )
+    .filter(
+      (elt) =>
+        !term ||
+        elt.firstName.toLowerCase().includes(term.toLowerCase()) ||
+        elt.surname.toLowerCase().includes(term.toLowerCase()) ||
+        elt.email.toLowerCase().includes(term.toLowerCase())
+    );
+
+  if (fltBookings.length <= 0) {
+    (term || date) && response.status(400).send("No matching results");
+  }
+
+  if (!term && !date) {
+    response.status(400).send("Please enter search term");
+  }
+
+  response.send(fltBookings);
+});
+
+// Get one booking by id
+app.get("/bookings/:id", function (req, res) {
+  let bookingIndex = bookings.findIndex((elt) => elt.id == request.params.id);
+
+  isInvalidId(request.params.id, bookingIndex, response);
+
+  if (bookingIndex >= 0) {
```

Comment:
This condition has already been checked in the `isInvalidId` function.

However there is also the case when `bookingIndex` is greater than the number of bookings. In this case you should send 404 Not Found status.